pkg/controller/direct/discoveryengine/datastore_controller.go

```diff
@@ -0,0 +1,280 @@
+	log := klog.FromContext(ctx)
+	if strings.HasPrefix(url, "//discoveryengine.googleapis.com/") {
+		id, err := krm.ParseDiscoveryEngineDataStoreExternal(url)
+		if err != nil {
+			log.V(2).Error(err, "url did not match DiscoveryEngineDataStore format", "url", url)
+		} else {
+			return &dataStoreAdapter{
+				model: m,
+				id:    id,
+			}, nil
+		}
+	}
+	return nil, nil
+}
+
+type dataStoreAdapter struct {
+	model   *dataStoreModel
+	id      *krm.DiscoveryEngineDataStoreID
+	desired *pb.DataStore
+	actual  *pb.DataStore
+}
+
+var _ directbase.Adapter = &dataStoreAdapter{}
+
+func (a *dataStoreAdapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("getting discoveryengine datastore", "name", a.id)
+
+	gcpClient, err := a.model.client(ctx, a.id.ProjectID)
+	if err != nil {
+		return false, err
+	}
+	req := &pb.GetDataStoreRequest{Name: a.id.String()}
+	actual, err := gcpClient.GetDataStore(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting discoveryengine datastore %q from gcp: %w", a.id.String(), err)
+	}
+
+	a.actual = actual
+	return true, nil
+}
+
+func (a *dataStoreAdapter) Create(ctx context.Context, createOp *directbase.CreateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("creating discoveryengine datastore", "name", a.id)
+
+	desired := direct.ProtoClone(a.desired)
+	desired.Name = a.id.String()
+
+	gcpClient, err := a.model.client(ctx, a.id.ProjectID)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.CreateDataStoreRequest{
+		Parent:      a.id.CollectionLink.String(),
+		DataStore:   desired,
+		DataStoreId: a.id.DataStore,
+	}
+	op, err := gcpClient.CreateDataStore(ctx, req)
+	if err != nil {
+		return fmt.Errorf("creating discoveryengine datastore %s: %w", a.id.String(), err)
+	}
+	created, err := op.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("discoveryengine datastore %s waiting creation: %w", a.id, err)
+	}
+	log.V(2).Info("successfully created discoveryengine datastore in gcp", "name", a.id)
+
+	status := &krm.DiscoveryEngineDataStoreStatus{}
+	mapCtx := &direct.MapContext{}
+	status.ObservedState = DiscoveryEngineDataStoreObservedState_FromProto(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	status.ExternalRef = direct.PtrTo(a.id.String())
+	return createOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *dataStoreAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating discoveryengine datastore", "name", a.id)
+
+	gcpClient, err := a.model.client(ctx, a.id.ProjectID)
+	if err != nil {
+		return err
+	}
+
+	desired := direct.ProtoClone(a.desired)
+	desired.Name = a.id.String()
+
+	// TODO(user): Update the field if applicable.
+	updateMask := &fieldmaskpb.FieldMask{}
+	if !reflect.DeepEqual(a.desired.DisplayName, a.actual.DisplayName) {
```

Comment:
Is `displayname` the only mutable field?

Reply:
I don't really know ... it's the only supported mutable field right now, I guess.  I do think we need more consistent treatment of mutability.